src/middleware.js

```diff
@@ -8,6 +8,24 @@ import { parseCid, tryParseCid } from './util/cid.js'
 
 const CF_CACHE_MAX_OBJECT_SIZE = 512 * Math.pow(1024, 2) // 512MB to bytes
 
+/**
+ * Creates a fresh context object that can be mutated by the request.
+ *
+ * The original context object is _shared_ among requests so should be cloned
+ * ASAP using this middleware and the clone should be mutated not the original.
+ *
+ * Some properties in the original context object are not enumerable so need
+ * to be expicitly added.
+ *
+ * @type {import('./bindings').Middleware<Context>}
+ */
+export function withContext (handler) {
+  return (request, env, ctx) => {
+    const context = { ...ctx, waitUntil: ctx.waitUntil.bind(ctx) }
```

Comment:
spread operator will only take care of top level elements in a nested object. Is there any risk we might incur with this?

Reply:
I almost did not do that because really there should be nothing else here to spread.

I think it's fine for stuff we've added (of which there shouldn't be anything anyway because of the risk it's changed by concurrent requests) and I'm pretty sure workers runtime wouldn't add anything to the original context that could change or be changed between requests since they know that the object is shared.